mike-petrut-2021-04-ndvi-automation.py

```diff
@@ -0,0 +1,743 @@
+# * Use the functions that we demonstrated during class to make your workflow more efficient.
+# * BONUS - if you  chose - you can export your data as a csv file. You will get bonus points for doing this.
+# 
+# 
+# ## Assignment Requirements
+# 
+# Your submission to the GitHub repository should include:
+# * This Jupyter Notebook file (.ipynb) with:
+#     * The code to create a plot of mean NDVI across a year for  2 NEON Field Sites:
+#         * NDVI on the x axis and formatted dates on the y for both NEON sites on one figure/axis object
+#     * The **data should be cleaned to remove the influence of clouds**. See the [earthdatascience website for an example of what your plot might look like with and without removal of clouds](https://www.earthdatascience.org/courses/earth-analytics-python/create-efficient-data-workflows/).
+# * BONUS: Create one output `.csv` file that has 3 columns - NDVI, Date and Site Name - with values for SJER and HARV.
+# 
+# Your notebook should:
+# * Have *at least* 2 well documented and well named functions with docstrings.
+# * Include a Markdown cell at the top of the notebook that outlines the overall workflow using pseudocode (i.e. plain language, not code)
+# * Include additional Markdown cells throughout the notebook to describe: 
+#     * the data that you used - and where it is from
+#     * how data are being processing
+#     * how the code is optimized to run fast and be more concise
+
+# # Replace this cell with your pseudocode  for this workflow
+# 
+# If you happen to be a diagram person a diagram is ok too
+# 
+# 
+***Workflow for Mean NDVI Modelling***
+
+1. Get list of all directories and associated site names.
+
+2. Open each site directory, get Landsat scenes, and calculate mean NDVI for each scene for that site.
+   # Steps for calculating the mean NDVI for each Landsat Scene
+   1. Create a loop that reads the site and date from each directory
+   2. Sort the bands by number 
+   3. Crop to the extent of the focus shapefile
+   4. Mask to the range of Landsat Values 
+   5. Optionally clean the image using the pixel cloud cover image 
+   
+3. Capture results (including mean NDVI, date, and site name) to a list or dataframe.
+   # Steps for sorting the table
+   1. append all values from the looped tasks in Step 2
+   2. Create pandas data frame, setting date as index 
+   
+4. Export dataframe with mean NDVI values to csv.
+# In[1]:
+
+
+# Autograding imports - do not modify this cell
+import matplotcheck.autograde as ag
+import matplotcheck.notebook as nb
+import matplotcheck.timeseries as ts
+from datetime import datetime
+
+
+# In[2]:
+
+
+# Import necessary packages
+import os
+from glob import glob
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+from shapely.geometry import box
+import geopandas as gpd
+import xarray as xr
```

Comment:
The order of the imports seems correct (besides datetime and matplotlib.dates at the end), but you could use spacing between first, second, and third party imports)